### ISSUE TITLE
[Enhancement]Show and limit the space used

### ISSUE THREAD

**Issue by BobAnkh** [LycheeOrg/Lychee#520]
*Explain in a few words which functionality or improvements you would like to see in Lychee.*

I really like to use Lychee for its simple.

I would like to have new features related to **space used**:

1. I would like to see how much space I have used to store the pictures uploaded so I can delete some pictures if I think I have used too much space. I have no idea of how much space I have used in the current version.
2. Admin user could have a view of how much space every user has used and the total space used so he can inform the users who occupy too much space.
3. Admin user could set different space limit to different users for better space schedule.

Thank you for considering my suggestions!

**Comment by d7415** [comment]
>     1. I would like to see how much space I have used to store the pictures uploaded so I can delete some pictures if I think I have used too much space. I have no idea of how much space I have used in the current version.

This is on the Diagnostics page :) 

Per-user stats would be more complicated but possible. All users' photos are stored in the same folders. At some point I seem to remember storing file sizes in the photos table, which would be required to make this manageable. I can't remember whether this is currently the case though.  

I think storage quotas are out of scope, though could be possible if file sizes were in the database as above.

**Comment by BobAnkh** [comment]
Thank you for your reply.

I'm sorry that I forget to check the Diagnostics page. So this is the total storage used by lychee which can be viewed by the admin user, right?

The features now are good enough for me to use it on my own. 

And I would like to share it with a group of friends, so I'm thinking of some features related to storage use to **schedule the total storage**. (Try to **avoid** some users to **overuse the total storage**)

Cause I'm also using Nextcloud on the same server which has the storage quotas for different users and it is very convenient for me to organize, so I'm thinking of lychee to have this feature for better storage scheduled.  Though pictures are stored in the same folders, I believe they can be distinguished easily and every picture shows its size in `About`, so I think checking the total size for different users is a reachable feature, maybe calculated every time when asked? 

(Have you ever thought about put different users' pictures into different folders? Cause all the pictures are in same folders, so I cannot come up with some external ways to limit different users' storage)

Thank you!

**Comment by ildyria** [comment]
The reason why the _size_ is "on request" in the Diagnostic page (while it was directly available there before) is due to the length of time required to compute it...

This is on my server:
![2020-04-30_1920x1080_12:06:22](https://user-images.githubusercontent.com/627094/80698889-65496080-8adb-11ea-9a5d-7bd6c6e17a85.png)

As you can see, I use about 110 GB for Lychee  (12 000+ pictures). Computing the space used requires basically 2 calls to getSize. The first one starts the computation (awk etc), the second call is using the cache...

As for adding a the size info in a table, there would be 2 possibilities:
- add 7 more columns (big, medium x2, small x2, thumbs x2) to the current Photo table.
- create a new table with those same information.
Also to be noted that the migration process might be long dues to the amount of pictures to process...

I am not opposed to adding this table quite the contrary as this would significantly speed up the space computation process and would add the requested feature above.

However at the time of this writing, I no longer have resources to dedicate to LycheeOrg for the next few months (currently writing my PhD thesis), so I can only hope that someone will make a PR for that.

On a side note, a simpler and quicker way to add a quota is simply to implement a limit to the number of pictures a user can upload and add it to a [middleware](https://github.com/LycheeOrg/Lychee/tree/master/app/Http/Middleware).

**Comment by BobAnkh** [comment]
I'm not familiar with this project and this field but I would like to ask that if it is possible to simply record the space used and simply show it whenever asked?

For example, just calculate the space used by lychee after installed or migrated, and when a new user is created the space used of him is simply recorded as zero. Then when a photo is uploaded or  deleted, just add or minus the space corresponding to the photo and refresh the usage.

I have no idea whether it is feasible or not, or maybe it is similar to what have been proposed. I just share my thoughts so please forgive if there are some mistakes.

**Comment by ildyria** [comment]
>'m not familiar with this project and this field but I would like to ask that if it is possible to simply record the space used and simply show it whenever asked?

At some point I was considering adding a space to the Cache infrastructure available, but that implied invalidating it as soon as you upload a picture...

The update part is a bit anoying.

Going back to the idea of table for size above, there is one thing to also consider is the fact that some pictures have the same url (that is what happen when you duplicate a picture). As a result, recording the size for each picture in each row would yield to a wrong result (for the sum).

>For example, just calculate the space used by lychee after installed or migrated, and when a new user is created the space used of him is simply recorded as zero. Then when a photo is uploaded or deleted, just add or minus the space corresponding to the photo and refresh the usage.

The idea of having size column in the user table is attractive but this might also get some tricky situtations when the admin is moving the picture from album of user A to album of user B...
Basically this implies space management in a lot of the base actions.

> I have no idea whether it is feasible or not, or maybe it is similar to what have been proposed. I just share my thoughts so please forgive if there are some mistakes.

It is feasible, but the question is more, is it the best solution?

**Comment by BobAnkh** [comment]
> when the admin is moving the picture from album of user A to album of user B...

As you have mentioned this, I would like to ask a question not related to this issue: 

Is the admin user can figure out which picture belongs which user in the large albums(cause I haven't found this). If so (or even the pictures can be classified according to users), I think manually management is acceptable to me, though a little bit trouble. At least it gives me a way to manage the storage temporarily as an alternative.

(Maybe different users get different folders will be easier to manage? Calculate a folder's size may be easier and quicker?)

If lychee has a better way implemented to manage, please tell me.

Thank you for your reply.

**Comment by ildyria** [comment]
> (Maybe different users get different folders will be easier to manage? Calculate a folder's size may be easier and quicker?)

That is the way we do it.
![2020-04-30_1920x1080_15:46:49](https://user-images.githubusercontent.com/627094/80718126-0a733180-8afa-11ea-92f1-6d25897bc31c.png)

But ownership is also per picture (required for the *unsorted* ones).

**Comment by BobAnkh** [comment]
Sorry, why I haven't seen `owner`? Is there something I've to configure?
![image](https://user-images.githubusercontent.com/44333669/80718956-64411f00-8b2d-11ea-80b0-6462f0103569.png)

**Comment by ildyria** [comment]
You are looking at a picture. Not an album. :)

**Comment by BobAnkh** [comment]
Sorry I have just searched in smart albums. It seems unsorted albums don't have this. (Users who don't create albums may be hard to be managed?) Maybe creating a album which has the same name of its user when a new user is created?

Anyway, at least there is a temporarily way to manage.

Really thank you for your help!
